### PR TITLE
Codecov integration

### PR DESCRIPTION
**Pull request by zuiris**
### Description of the Change

Added coverage calculation and automatic coverage uploading to codecov

### Benefits
This is a common practice and coverage analysis will be required to establish quality gate in the future

### Possible Drawbacks 
Additional step in the pipeline -> it will take more time for PRs to be processed by CI

### Usage Examples or Tests *[optional]*
None 

### Alternate Designs *[optional]*
None

### CODE REVIEW

File: .github/workflows/ubsan.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        GITHUB_HUNTER_TOKEN: ${{ secrets.GITHUB_HUNTER_TOKEN }}
+        CFLAGS: "-I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"
+        CXXFLAGS: "-isysroot/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"
+      run: cmake . -GNinja -Bbuild -D UBSAN=ON
+    - name: build
+      run: cmake --build build -- -j2
+    - name: run tests
+      env:
+        CTEST_OUTPUT_ON_FAILURE: 1
+      run: cmake --build build --target test
```
Comment (Alexey-N-Chernyshov) [inline .github/workflows/ubsan.yml]: Please, add newline at end of file.

File: .github/workflows/tsan.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        GITHUB_HUNTER_TOKEN: ${{ secrets.GITHUB_HUNTER_TOKEN }}
+        CFLAGS: "-I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"
+        CXXFLAGS: "-isysroot/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"
+      run: cmake . -GNinja -Bbuild -D TSAN=ON
+    - name: build
+      run: cmake --build build -- -j2
+    - name: run tests
+      env:
+        CTEST_OUTPUT_ON_FAILURE: 1
+      run: cmake --build build --target test
```
Comment (Alexey-N-Chernyshov) [inline .github/workflows/tsan.yml]: Please, add newline at end of file.

File: .github/workflows/msan.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} # has to be included to access other secrets
+        GITHUB_HUNTER_USERNAME: ${{ secrets.GITHUB_HUNTER_USERNAME }}
+        GITHUB_HUNTER_TOKEN: ${{ secrets.GITHUB_HUNTER_TOKEN }}
+      run: cmake . -GNinja -Bbuild -D MSAN=ON
+    - name: build
+      run: cmake --build build -- -j2
+    - name: run tests
+      env:
+        CTEST_OUTPUT_ON_FAILURE: 1
+      run: cmake --build build --target test
```
Comment (Alexey-N-Chernyshov) [inline .github/workflows/msan.yml]: Please, add newline at end of file.

File: .github/workflows/lsan.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} # has to be included to access other secrets
+        GITHUB_HUNTER_USERNAME: ${{ secrets.GITHUB_HUNTER_USERNAME }}
+        GITHUB_HUNTER_TOKEN: ${{ secrets.GITHUB_HUNTER_TOKEN }}
+      run: cmake . -GNinja -Bbuild -D LSAN=ON
+    - name: build
+      run: cmake --build build -- -j2
+    - name: run tests
+      env:
+        CTEST_OUTPUT_ON_FAILURE: 1
+      run: cmake --build build --target test
```
Comment (Alexey-N-Chernyshov) [inline .github/workflows/lsan.yml]: Please, add newline at end of file.

File: .github/workflows/asan.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        CFLAGS: "-I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"
+        CXXFLAGS: "-isysroot/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"
+        ASAN_OPTIONS: "halt_on_error=0"
+      run: cmake . -GNinja -Bbuild -D ASAN=ON
+    - name: build
+      run: cmake --build build -- -j2
+    - name: run tests
+      env:
+        CTEST_OUTPUT_ON_FAILURE: 1
+      run: cmake --build build --target test
```
Comment (Alexey-N-Chernyshov) [inline .github/workflows/asan.yml]: Please, add newline at end of file.

File: _3rdParty/Hunter/toolchain/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+list(REMOVE_DUPLICATES macros_list)
+string(REPLACE ";" "\n" macros_string "${macros_list}")
+
+string(COMPARE EQUAL "${macros_string}" "" is_empty)
+if(is_empty)
+  hunter_fatal_error("No toolchain info generated" ERROR_PAGE error.no.toolchain.info)
+endif()
+
+file(APPEND "${TOOLCHAIN_INFO_FILE}" "Predefined macroses:\n${macros_string}")
```
Comment (Alexey-N-Chernyshov) [inline _3rdParty/Hunter/toolchain/CMakeLists.txt]: "Macros" is a plural form itself.

File: _3rdParty/Hunter/toolchain/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,153 @@
+cmake_minimum_required(VERSION 3.0)
```
Comment (igor-egorov) [inline _3rdParty/Hunter/toolchain/CMakeLists.txt]: Should this file be removed?

**Comment by igor-egorov** [comment]
Also, please consider disabling ASAN and MSAN till the project gets fixed

**Pull request merged**